File: windows/Tokendial.Tests/ForecastTests.cs
using Tokendial.Core.Model;

namespace Tokendial.Tests;

public class ForecastTests
{
    private static readonly DateTimeOffset T0 = new(2026, 9, 8, 14, 0, 0, TimeSpan.Zero);

    private static List<UsageSample> Ramp(double startFraction, double perMinute, int minutes, int step = 1, DateTimeOffset? resetsAt = null)
    {
        var samples = new List<UsageSample>();
        for (var m = 0; m <= minutes; m += step) samples.Add(new UsageSample(T0.AddMinutes(m), startFraction + perMinute * m, resetsAt));
        return samples;
    }

    [Fact]
    public void ASteadyClimbRunsOutAtTheExpectedTime()
    {
        var reset = T0.AddHours(3);
        var samples = Ramp(0.40, 0.01, 20, resetsAt: reset);
        var forecast = Forecast.For(samples, T0.AddMinutes(20), reset);
        Assert.NotNull(forecast);
        Assert.Equal(ForecastKind.RunsOut, forecast.Kind);
        Assert.Equal(T0.AddMinutes(60), forecast.RunsOutAt!.Value, TimeSpan.FromSeconds(30));
    }

    [Fact]
    public void ASlowClimbLastsUntilTheReset()
    {
        var reset = T0.AddMinutes(40);
        var samples = Ramp(0.40, 0.01, 20, resetsAt: reset);
        var forecast = Forecast.For(samples, T0.AddMinutes(20), reset);
        Assert.Equal(ForecastKind.LastsUntilReset, forecast?.Kind);
        Assert.Null(forecast?.RunsOutAt);
    }

    [Fact]
    public void AFlatWindowHasNoForecast()
    {
        var samples = Ramp(0.40, 0, 30, step: 5);
        Assert.Null(Forecast.For(samples, T0.AddMinutes(30), T0.AddHours(2)));
    }

    [Fact]
    public void TooFewOrTooCloseSamplesHaveNoForecast()
    {
        Assert.Null(Forecast.For(Ramp(0.40, 0.01, 2), T0.AddMinutes(2), T0.AddHours(2)));
        Assert.Null(Forecast.For(Ramp(0.40, 0.01, 8), T0.AddMinutes(8), T0.AddHours(2)));
    }

    [Fact]
    public void OldSamplesAreIgnored()
    {
        var samples = Ramp(0.10, 0.01, 30);
        Assert.Null(Forecast.For(samples, T0.AddMinutes(30 + 61), T0.AddHours(3)));
    }

    [Fact]
    public void AFullWindowHasNoForecast()
    {
        var samples = Ramp(0.80, 0.01, 20);
        Assert.Null(Forecast.For(samples, T0.AddMinutes(20), T0.AddHours(3)));
    }

    [Fact]
    public void WithoutAResetOnlyTheNextDayCounts()
    {
        Assert.Equal(ForecastKind.RunsOut, Forecast.For(Ramp(0.40, 0.01, 20), T0.AddMinutes(20), null)?.Kind);
        Assert.Null(Forecast.For(Ramp(0.40, 0.0001, 20), T0.AddMinutes(20), null));
    }

    [Fact]
    public void IntegerPercentagesStillGiveASlope()
    {
        var samples = new List<UsageSample>();
        for (var m = 0; m <= 20; m++) samples.Add(new UsageSample(T0.AddMinutes(m), Math.Floor((0.40 + 0.01 * m) * 100) / 100, null));
        var slope = Forecast.SlopePerHour(samples);
        Assert.InRange(slope, 0.55, 0.65);
    }
}

public class UsageHistoryTests
{
    private static readonly DateTimeOffset T0 = new(2026, 9, 8, 14, 0, 0, TimeSpan.Zero);

    [Fact]
    public void KeepsSamplesOfOneEpochAndTrimsByAgeAndCount()
    {
        var history = new UsageHistory();
        for (var m = 0; m < 400; m++) history.Add("claude", "session", new UsageSample(T0.AddMinutes(m), Math.Min(0.99, m / 500.0), T0.AddHours(8)));
        var samples = history.Samples("claude", "session");
        Assert.True(samples.Count <= UsageHistory.MaxSamples);
        Assert.True(samples.All(s => T0.AddMinutes(399) - s.TakenAt <= UsageHistory.MaxAge));
    }

    [Fact]
    public void AMovedResetStartsANewEpoch()
    {
        var history = new UsageHistory();
        history.Add("claude", "session", new UsageSample(T0, 0.9, T0.AddHours(1)));
        history.Add("claude", "session", new UsageSample(T0.AddMinutes(1), 0.95, T0.AddHours(1).AddSeconds(30)));
        Assert.Equal(2, history.Samples("claude", "session").Count);
        history.Add("claude", "session", new UsageSample(T0.AddMinutes(2), 0.02, T0.AddHours(6)));
        Assert.Single(history.Samples("claude", "session"));
    }

    [Fact]
    public void ADropWithoutAResetStartsANewEpoch()
    {
        var history = new UsageHistory();
        history.Add("antigravity", "requests", new UsageSample(T0, 0.6, null));
        history.Add("antigravity", "requests", new UsageSample(T0.AddMinutes(1), 0.55, null));
        Assert.Equal(2, history.Samples("antigravity", "requests").Count);
        history.Add("antigravity", "requests", new UsageSample(T0.AddMinutes(2), 0.1, null));
        Assert.Single(history.Samples("antigravity", "requests"));
    }

    [Fact]
    public void RecordsOnlyWindowsWithAFraction()
    {
        var history = new UsageHistory();
        var reading = new ProviderReading("codex", "Codex", Fidelity.Official, ReadingStatus.LiveNow,
            [new UsageWindow("primary", "5h limit", 0.3, ResetsAt: T0.AddHours(2)), new UsageWindow("count", "Requests", Count: 7)]);
        history.Record(reading, T0);
        Assert.Single(history.Samples("codex", "primary"));
        Assert.Empty(history.Samples("codex", "count"));
        history.Forget("codex");
        Assert.Empty(history.Samples("codex", "primary"));
    }
}

File: macos/TokendialCore/Tests/TokendialCoreTests/ForecastTests.swift
import XCTest
@testable import TokendialCore

private let t0 = Date(timeIntervalSince1970: 1_788_616_800) // 2026-09-08 14:00 UTC

private func at(_ minutes: Double) -> Date { t0.addingTimeInterval(minutes * 60) }

private func ramp(_ start: Double, _ perMinute: Double, _ minutes: Int, step: Int = 1, resetsAt: Date? = nil) -> [UsageSample] {
    stride(from: 0, through: minutes, by: step).map {
        UsageSample(takenAt: at(Double($0)), usedFraction: start + perMinute * Double($0), resetsAt: resetsAt)
    }
}

final class ForecastTests: XCTestCase {
    func testASteadyClimbRunsOutAtTheExpectedTime() throws {
        let reset = at(180)
        let forecast = try XCTUnwrap(Forecast.of(ramp(0.40, 0.01, 20, resetsAt: reset), now: at(20), resetsAt: reset))
        XCTAssertEqual(forecast.kind, .runsOut)
        let runsOut = try XCTUnwrap(forecast.runsOutAt)
        XCTAssertEqual(runsOut.timeIntervalSince(at(60)), 0, accuracy: 30)
    }

    func testASlowClimbLastsUntilTheReset() {
        let reset = at(40)
        let forecast = Forecast.of(ramp(0.40, 0.01, 20, resetsAt: reset), now: at(20), resetsAt: reset)
        XCTAssertEqual(forecast?.kind, .lastsUntilReset)
        XCTAssertNil(forecast?.runsOutAt)
    }

    func testAFlatWindowHasNoForecast() {
        XCTAssertNil(Forecast.of(ramp(0.40, 0, 30, step: 5), now: at(30), resetsAt: at(120)))
    }

    func testTooFewOrTooCloseSamplesHaveNoForecast() {
        XCTAssertNil(Forecast.of(ramp(0.40, 0.01, 2), now: at(2), resetsAt: at(120)))
        XCTAssertNil(Forecast.of(ramp(0.40, 0.01, 8), now: at(8), resetsAt: at(120)))
    }

    func testOldSamplesAreIgnored() {
        XCTAssertNil(Forecast.of(ramp(0.10, 0.01, 30), now: at(91), resetsAt: at(180)))
    }

    func testAFullWindowHasNoForecast() {
        XCTAssertNil(Forecast.of(ramp(0.80, 0.01, 20), now: at(20), resetsAt: at(180)))
    }

    func testWithoutAResetOnlyTheNextDayCounts() {
        XCTAssertEqual(Forecast.of(ramp(0.40, 0.01, 20), now: at(20), resetsAt: nil)?.kind, .runsOut)
        XCTAssertNil(Forecast.of(ramp(0.40, 0.0001, 20), now: at(20), resetsAt: nil))
    }

    /// Vendors publish whole percentages, so neighbouring samples read flat; the fit still finds the pace.
    func testIntegerPercentagesStillGiveASlope() {
        let samples = (0...20).map { m in
            UsageSample(takenAt: at(Double(m)), usedFraction: ((0.40 + 0.01 * Double(m)) * 100).rounded(.down) / 100, resetsAt: nil)
        }
        let slope = Forecast.slopePerHour(samples)
        XCTAssertGreaterThan(slope, 0.55)
        XCTAssertLessThan(slope, 0.65)
    }
}

final class UsageHistoryTests: XCTestCase {
    func testKeepsSamplesOfOneEpochAndTrimsByAgeAndCount() {
        let history = UsageHistory()
        for m in 0..<400 {
            history.add("claude", "session", UsageSample(takenAt: at(Double(m)), usedFraction: min(0.99, Double(m) / 500), resetsAt: at(480)))
        }
        let samples = history.samples("claude", "session")
        XCTAssertLessThanOrEqual(samples.count, UsageHistory.maxSamples)
        XCTAssertTrue(samples.allSatisfy { at(399).timeIntervalSince($0.takenAt) <= UsageHistory.maxAge })
    }

    func testAMovedResetStartsANewEpoch() {
        let history = UsageHistory()
        history.add("claude", "session", UsageSample(takenAt: t0, usedFraction: 0.9, resetsAt: at(60)))
        history.add("claude", "session", UsageSample(takenAt: at(1), usedFraction: 0.95, resetsAt: at(60).addingTimeInterval(30)))
        XCTAssertEqual(history.samples("claude", "session").count, 2)
        history.add("claude", "session", UsageSample(takenAt: at(2), usedFraction: 0.02, resetsAt: at(360)))
        XCTAssertEqual(history.samples("claude", "session").count, 1)
    }

    func testADropWithoutAResetStartsANewEpoch() {
        let history = UsageHistory()
        history.add("antigravity", "requests", UsageSample(takenAt: t0, usedFraction: 0.6, resetsAt: nil))
        history.add("antigravity", "requests", UsageSample(takenAt: at(1), usedFraction: 0.55, resetsAt: nil))
        XCTAssertEqual(history.samples("antigravity", "requests").count, 2)
        history.add("antigravity", "requests", UsageSample(takenAt: at(2), usedFraction: 0.1, resetsAt: nil))
        XCTAssertEqual(history.samples("antigravity", "requests").count, 1)
    }

    func testRecordsOnlyWindowsWithAFraction() {
        let history = UsageHistory()
        let reading = ProviderReading(providerId: "codex", displayName: "Codex", fidelity: .official, status: .live,
                                      windows: [UsageWindow(id: "primary", label: "5h limit", usedFraction: 0.3, resetsAt: at(120)),
                                                UsageWindow(id: "count", label: "Requests", count: 7)])
        history.record(reading, now: t0)
        XCTAssertEqual(history.samples("codex", "primary").count, 1)
        XCTAssertTrue(history.samples("codex", "count").isEmpty)
        history.forget("codex")
        XCTAssertTrue(history.samples("codex", "primary").isEmpty)
    }
}
